### ISSUE TITLE
Define Metrics for Debugability

### ISSUE THREAD

**Issue by AMZN-scspaldi** [o3de/sig-testing#47]
<html><body>
<h2 id="RFCDefineMetricsforDebugability-Summary">Summary</h2><p>Currently,
 most design decisions and work for debugability are done through either
 grass-roots movements where small groups of developers make a change, <span class="inline-comment-marker" data-ref="fbd29f9f-de41-4973-b604-ada427d4e0a8">or a contributor pushes</span>
 for a specific improvement. It is consequently unclear what metrics 
have been important in past decisions. We want to design and implement 
metrics so that such decisions can be supported by appropriate data and 
to measure the effectiveness of changes made because of these decisions.</p><p>Selecting
 and measuring appropriate metrics for debugability is challenging. 
Ultimately, debugability has a significant human element that is 
difficult to quantify. However, it is still valuable in measuring what 
can be measured, as well as acknowledging the incomplete picture that 
these metrics provide.</p><p>There is also a wide range of improvements 
that can be done for debugability timelines that cross over into other 
areas. For example, faster build times also improves debugability 
timelines. However, to limit the scope of this RFC, we are going to limit this document to discussing elements that 
are unique to debugability.</p><p>We are suggesting to measure a few key debugging times and polling customers on their debugging experience.</p><p>
                </span>

</p><h2 id="RFCDefineMetricsforDebugability-Whatisthemotivationforthissuggestion?">What is the <span class="inline-comment-marker" data-ref="e1f48d76-33b9-4e44-9c9c-19ec461b8179">motivation</span> for this suggestion?</h2><p>Primarily,
 the motivation for this is to enable metrics-driven decisions around 
improvements to debugability. Diving a little deeper, these metrics will
 allow us to more intelligently focus our efforts in improving 
debugability, specifically in improving the timelines for discovering 
and fixing bugs. It will also allow us to prioritize which problems are 
most severe.<br><br><span class="inline-comment-marker" data-ref="8f4d26ec-21ae-4c23-aa6b-21022b04984e">We
 also want to be able to track the impact of decisions we make to 
improve debugability. Measuring metrics and their subsequent improvement
 is one way to do this. This will allow us to provide visibility to the 
larger organization by a metrics-driven approach and provide support for
 claims of debugability improvements.</span></p><h2 id="RFCDefineMetricsforDebugability-DesignDescription">Design Description</h2><h3 id="RFCDefineMetricsforDebugability-DebuggingSteps">Debugging Steps</h3><p>Here are some common steps of finding and fixing a bug:</p>

<ol>
  <li>The error is reported. This is done by Jenkins sending out an e-mail, although developers can also view the Jenkins build through its web UI.</li>
  <li>The log is investigated. This can be found through the web UI for Jenkins. (or in a GitHub issue)
    <ol>
      <li>This may include artifact investigation, which is also found through the web UI for Jenkins</li>
    </ol>
  </li>
<li>The bug is reproduced locally. This step is not always required, but often is. Depending on the developer's workstation state, the nature of the bug, and the developer's preferences, the following sub-steps may be required:
  <ol>
    <li>Sync the commit where the bug was found.</li>
    <li>Build the commit in the appropriate flavor.</li>
    <li>Add breakpoints. This is a less common step.</li>
    <li>Add new output. This is a less common step.</li>
    <li>Run the appropriate executable(s).</li>
    <li>Run the appropriate test(s).</li>
  </ol>
</li>
<li>Make changes to actually fix the bug.
  <ol>
    <li>If no test existed to catch the bug, add an appropriate test(s) to prevent the bug from regressing.</li>
  </ol>
</li>
</ol>

<h3 id="RFCDefineMetricsforDebugability-MeasuringDebuggingTime"><span class="inline-comment-marker" data-ref="59a45deb-1295-469e-ad6f-86ceecaf73c6">Measuring Debugging Time</span></h3><p>Each of the steps above <span class="inline-comment-marker" data-ref="e0fe9fcf-00f3-43db-877f-199c0912f0ee">can be measured in time</span>,
 although the value of those data points will vary wildly. Many of them 
are too dependent on the bug, can be done in parallel, or are out of 
scope for debugability. In order to ensure consistent measuring, we will
 perform a UX study where research subjects are monitored by a 
researcher. The researcher will measure the time each step takes with a 
stopwatch, record the results, and collect other feedback.</p><p>Local 
reproduction is a borderline case - things like build time are out of 
scope, but making sure that the bug can be reproduced locally reliably 
are in scope. Further, there's a ton of variables involved that will 
make any metrics we do measure exceptionally noisy. Did they have a 
partial sync already? Did they have a build already? Do they prefer to 
debug via logs or debug via the debugger? All of these will cause the 
time to vary wildly while not actually indicating the ease of the 
debugging as defined as in-scope for this RFC. 
Given that local reproduction is more of a binary question (Can we 
reproduce the bug locally?) given the scope of this RFC, we will not be 
including it as a suggested timed metric.</p><h3 id="RFCDefineMetricsforDebugability-DebuggingTimeMetrics">Debugging Time Metrics</h3><p><span class="inline-comment-marker" data-ref="60f9f581-acb7-4d2b-bea5-a8cb7332f1eb">Thus, this leaves us with </span><span class="inline-comment-marker" data-ref="1839efeb-f377-4499-907a-5311d4e1cd93"><span class="inline-comment-marker" data-ref="60f9f581-acb7-4d2b-bea5-a8cb7332f1eb">3 steps</span></span><span class="inline-comment-marker" data-ref="60f9f581-acb7-4d2b-bea5-a8cb7332f1eb"> to measure for time:</span></p><ul><li><span class="inline-comment-marker" data-ref="c97c4ee4-2b41-44c0-badf-54ad98aead7f">Error reporting time</span>:
 This step would have three parts. Currently, Jenkins only outputs an 
error e-mail when all builds in a run are finished, but changes could be
 made to report an issue as soon as a build in a run has failed.<ul><li>Time when failure is first signaled inside Jenkins job</li><li>Time when Jenkins sends an e-mail reporting the failure to the user (or a bot posts a failure message)</li><li>Time when user takes action on the notification and starts working on their machine (modifying code, or starting a build)</li></ul></li><li><span class="inline-comment-marker" data-ref="09b3db6e-23ba-41b1-9046-b8976465f618">Log-finding time</span>:
 This step would have two parts. We expect this to be a small portion of
 the total time consumed, but it is still valuable to reduce it where 
possible.<ul><li>Time it takes from notification to finding the correct log</li><li>Time it takes (after finding the correct log) to parse enough information in the log to take action locally</li></ul></li><li>Artifact-finding
 time: We expect this to be a small portion of the total time consumed, 
but it is still valuable to reduce it where possible.</li></ul><p>The 
first one is a binary option - we can either send the e-mail at the end 
of the build, or upon the first failure. But measuring that difference 
could allow us to decide if the earlier messaging time is worthwhile. 
The second two options are going to have to be user-reported metrics.</p><div class="table-wrap">

<html><body>
<p><br /></p>

Metric | Description | Measured By
-- | -- | --
Failure first signaled | Time to see first failure in Jenkins | Automated script
Failure e-mail sent | How long it takes for an e-mail/message to go out after a failure is found | Automated script
Log-finding | How long it takes & how many clicks to find the appropriate log from a failed Jenkins run main page | Polling, customer reports
Log-parsing | How long it takes to find the actual error message in the appropriate log | Polling, customer reports
Artifact-finding | How long it takes & how many clicks to find artifacts from a failed Jenkins run main page | Polling, customer reports

</body>
</html>

## Steps to Achieve Goal

Even if a given objective is swift to complete, having too many steps (often in tedious clicks), can be a frustrating developer experience. Thus, we will measure the amount of steps a given goal took prior to changes and after changes. This metric has the advantage of not relying on polling, either, as we can independently measure this.

## Polling

While potentially less reliable than absolute metrics such as measuring time, polling customers for their preferences and pain points is still valuable. We would reach out to the sig-ux in order to develop questions. Measuring user preferences is not a primary skill set of our team and we would want to lean on experts to ensure an effective solution for our customers.

In addition to broad polls, we will include deeper individual customer reports where we work with a single developer to examine their experience directly. For this step, we would work with a small sample of specific developers to get a detailed view of their debugging experience before and after changes. While this will take more time, it will provide more reliable results and allow us to guarantee accuracy in our polling.

### Checks

As a part of polling, we want to ensure that certain key checks are hit. These are:

* Were you able to reproduce the bug locally, the first time?
* Were you able to reproduce the bug locally?
* Were you able to reproduce the bug?

These are, in order, valuable benchmarks to hit. Reproduction of a bug is key in fixing it. Being able to reproduce it locally allows a developer to glean important information in fixing it. Finally, being able to reproduce it the first time is helpful in simply saving developer time by not forcing them to repeat steps.

### Implementation

* Write script that scrapes Jenkins build logs to measure Failure first signaled and Failure e-mail sent metrics
* Perform external developer polling (this would be where most questions would be) (this would include log-finding, log-parsing, artifact-finding metrics)

<html><body>
<h3>Possible Outcomes / Scenarios</h3><p>Between metrics and polling, we will be able to generate a better picture of our pain sources and problems than we have now. While we won't be able to predict every possible outcome of these metrics and polling, we will attempt to identify several likely ones and comment briefly on our plans for those outcomes.</p>

Outcome/Scenario | Planned Steps
-- | --
Problems are not being logged due to unclear issue-filing process | Provide/improve documentation on issue-filing, run trainings on issue-filing
Problems are not being logged due to effort required | Provide quick-links and other shortcuts, create issue templates
Problems are being logged, but are not being actioned on | Use metrics to provide data to prioritize issues higher, reach out to appropriate owners

</body>
</html>

</div><h2 id="RFCDefineMetricsforDebugability-Whataretheadvantagesofthesuggestion?">What are the advantages of the suggestion?</h2><ul><li>Provides absolute metrics in debugability (see table above)</li><li>Polls customers directly to get their experiences</li><li>Focuses on the key aspects of debugability and provides a limited scope</li><li>Low investment cost</li></ul><h2 id="RFCDefineMetricsforDebugability-Whatarethedisadvantagesofthesuggestion?">What are the disadvantages of the suggestion?</h2><ul><li>Not all aspects are measured with metrics, meaning that some pain points may be under-examined</li><li>Metrics are only a partial picture of the problem, <span class="inline-comment-marker" data-ref="9a5bdfbb-3ae3-4dda-9a59-a12e4fc070b7">potentially supporting some solutions more than they merit (this can be mitigated by seeking customer feedback)<br></span></li><li>Customer opinions may change over time, <span class="inline-comment-marker" data-ref="4cae0660-1d36-4006-b482-3f5287e81250">requiring multiple polls (may require setting a particular cadence in order to maximize responses)<br></span></li><li>Some issues found by metrics/polling may not be relevant to the scope</li></ul><h2 id="RFCDefineMetricsforDebugability-Arethereanyalternativestothissuggestion?">Are there any alternatives to this suggestion?</h2><p>This
 suggestion has multiple components that are not coupled, so each 
component can be considered individually. This runs the risk of giving a
 partial picture, but perhaps the lesser amount of investment is worth 
the lesser result.</p><p>We could implement metrics on every step of the
 debugging process. This has a high risk of noise, as certain steps 
running long are way less impactful than others (builds can be run 
overnight, for example, while finding the logs &amp; artifacts is an 
active step). Further, many of these metrics can be addressed by tasks 
out of scope for this RFC.<br><br>We could skip the detailed customer 
experience poll. Broad polling often has low engagement, however, so a 
more direct interaction is expected to provide better results, even if 
it is at a higher cost.</p><h2 id="RFCDefineMetricsforDebugability-Whatisthestrategyforadoption?">What is the strategy for adoption?</h2><p>First, this would require reaching out to coordinate with <span class="inline-comment-marker" data-ref="a1c5999b-098b-42ec-bf8f-23feefb3e585">the sig-ux</span>
 to develop polling questions. Metrics and polling both require customer
 input, so this would require a customer outreach campaign. We would 
also have a small number of direct customer interactions. We would have 
to collate these results, then make our changes, then <span class="inline-comment-marker" data-ref="ec22c341-5ca4-4f6a-87cd-cbd0345435e6">poll again</span> to see if our <span class="inline-comment-marker" data-ref="71affdae-edd8-416e-83ee-b28eeef44a4b">changes have had the desired effects</span>.</p>
</body>
</html>

**Comment by Kadino** [comment]
@Kadino per discussion at the recent SIG-Testing meeting, move ready-for-review RFC to exist as a pull request (along with already reviewed RFCs)

**Comment by Kadino** [comment]
May be appropriate to drive this through only a user survey, instead of trying to set up UX experiments with time points.

**Comment by Kadino** [comment]
RFC exists at https://github.com/o3de/sig-testing/blob/main/rfcs/debuggability_metrics.md
